### PR TITLE
kubelet: new kubelet config option for disabling group oom kill

### PR DESCRIPTION
**Pull request by tzneal**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Added `singleProcessOOMKill` flag to the kubelet configuration. Setting that to true enable single process OOM killing in cgroups v2. In this mode, if a single process is OOM killed within a container, the remaining processes will not be OOM killed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e_node/oomkiller_linux_test.go [hunk 1]
```diff
@@ -150,6 +166,60 @@ func runOomKillerTest(f *framework.Framework, testCase testCase, kubeReservedMem
 	})
 }
 
+func runOomKillerFailsToKillTest(f *framework.Framework, testCase testCase, kubeReservedMemory float64) {
```
[the remark this answers is not in the document]
Comment (tzneal) [reply test/e2e_node/oomkiller_linux_test.go:169]: I think it is now, see the SIGDescribe line:

```
var _ = SIGDescribe("OOMKiller for pod using more memory than node allocatable [LinuxOnly]", framework.WithSerial(), func() {

```

[the remark this answers is not in the document]
Comment (haircommander) [reply test/e2e_node/oomkiller_linux_test.go:169]: ah yes correct! https://github.com/kubernetes/kubernetes/pull/122813/files#diff-61cad72537cf5e19d70deea7be51a6054f20cf41c15b3d949438edcb12506a40R50

File: test/e2e_node/oomkiller_linux_test.go [hunk 2]
```diff
@@ -150,6 +166,61 @@ func runOomKillerTest(f *framework.Framework, testCase testCase, kubeReservedMem
...
+					initialConfig.KubeReserved = map[string]string{}
+				}
+				// There's a race condition observed between system OOM and cgroup OOM if node alocatable
+				// memory is equal to node capacity. Hence, reserving a fraction of node's memory capacity for
+				// K8s components such that node allocatable memory is less than node capacity to
+				// observe OOM kills at cgroup level instead of system OOM kills.
+				initialConfig.KubeReserved["memory"] = fmt.Sprintf("%d", int(kubeReservedMemory*getLocalNode(context.TODO(), f).Status.Capacity.Memory().AsApproximateFloat64()))
+			})
+		}
+		if testCase.enableSingleProcessKill {
```
Comment (haircommander) [inline test/e2e_node/oomkiller_linux_test.go:188]: if we're only using this function when this is true, do we need to branch here? or: can we consolidate the two functions and branch on this variable?

**Comment by tzneal** [comment]
Also, I have no attachment to the name `SingleProcessOOMKill`, if anyone has a better name.

File: test/e2e_node/oomkiller_linux_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzneal) [reply test/e2e_node/oomkiller_linux_test.go:169]: @kannon92 is correct, I pointed you at the wrong SIGDescribe line 😄

File: test/e2e_node/oomkiller_linux_test.go (hunk 2 above)
Comment (tzneal) [reply test/e2e_node/oomkiller_linux_test.go:131]: Either is good with me, updated to branch in the test function.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go [hunk 3]
```diff
@@ -234,7 +235,7 @@ func (m *kubeGenericRuntimeManager) calculateLinuxResources(cpuRequest, cpuLimit
 	}
 
 	// runc requires cgroupv2 for unified mode
-	if isCgroup2UnifiedMode() {
+	if isCgroup2UnifiedMode() && !m.singleProcessOOMKill {
```
Comment (kannon92) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:238]: You should add some unit tests for this: https://github.com/kubernetes/kubernetes/blob/ee5e6e899b100edbd64860bc20337e11711e2670/pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go#L238

Test are all ready there so let's at least cover this case.

**Comment by kannon92** [comment]
I think we should make sure to document this option on the website also. Can you draft a PR while this is fresh?

I think putting this option https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-out-of-memory-behavior here would be appopriate.

@sftim WDYT?

**Comment by tzneal** [comment]
> I think we should make sure to document this option on the website also. Can you draft a PR while this is fresh?
> 
> I think putting this option https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-out-of-memory-behavior here would be appopriate.
> 
> @sftim WDYT?

I think it would have generated docs at https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1 . I'm not sure we want to document it further than that, its not something that's recommended more of a special case.

**Comment by tzneal** [comment]
> I think we should make sure to document this option on the website also. Can you draft a PR while this is fresh?
> 
> I think putting this option https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-out-of-memory-behavior here would be appopriate.
> 
> @sftim WDYT?

It should be auto-documented at https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1. I'm not sure if we want more docs than that, this isn't something to recommend.

**Comment by kannon92** [comment]
> Changelog suggestion
> 
> ```diff
> -Adds a new `singleProcessOOMKill` flag to the Kubelet config to enable single process killing in cgroupsv2. In this mode if a single process is killed within a container, the remaining processes will not be killed.
> +Added  `singleProcessOOMKill` flag to the kubelet configuration. Setting that to true enable single process killing in cgroups v2. In this mode, if a single process is killed within a container, the remaining processes will not be killed.
> ```
> 
> If this flag is false and I exec into a container and kill one of the processes there (but not container's PID 1), what happens? I'd expect that regardless of this flag's value, killing a process other than the init process leads to no special behavior.
> 
> What did I miss? The docs - that we don't yet have - aren't clear.

I tagged you to suggest if we should document this config option and if I linked a correct place to document this. But @tzneal points out that this feature is not recommended yet so maybe we can pass on a doc update.

**Comment by tzneal** [comment]
> this feature is not recommended yet so maybe we can pass on a doc update.

It's not that the feature isn't recommended yet, I don't think it will ever be recommended. With cgroupsv2, `memory.oom.group` is set by default which causes OOM kills to work like the documentation has stated (i.e. https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/ where the docs talk about containers being killed).  It didn't work that way previously due needing a user space handler to implement what `memory.oom.group` does automatically.

This change is just to add support for the hopefully small number of users that want the arguably broken behavior back where individual processes in a container can be OOM killed, but the remaining processes keep running.

**Comment by sftim** [comment]
> @tzneal points out that this feature is not recommended yet so maybe we can pass on a doc update.

For thing you could use but we don't recommend, readers of the website won't know we don't recommend it unless we mark it as alpha or otherwise tell 'em that we don't recommend using it.

If it's not behind an alpha feature gate, document it. At least enough that people know where they stand.

**Comment by tzneal** [comment]
> > @tzneal points out that this feature is not recommended yet so maybe we can pass on a doc update.
> 
> For thing you could use but we don't recommend, readers of the website won't know we don't recommend it unless we mark it as alpha or otherwise tell 'em that we don't recommend using it.
> 
> If it's not behind an alpha feature gate, document it. At least enough that people know where they stand.

Added https://github.com/kubernetes/website/pull/44781/

**Comment by haircommander** [comment]
> If this flag is false and I exec into a container and kill one of the processes there (but not container's PID 1), what happens? I'd expect that regardless of this flag's value, killing a process other than the init process leads to no special behavior.

good point. the doc should explitly say `OOM killed`, or maybe `OOM killed by the kernel` to clarify here.

**Comment by tzneal** [comment]
> > If this flag is false and I exec into a container and kill one of the processes there (but not container's PID 1), what happens? I'd expect that regardless of this flag's value, killing a process other than the init process leads to no special behavior.
> 
> good point. the doc should explitly say `OOM killed`, or maybe `OOM killed by the kernel` to clarify here.
Thanks, I updated the docs to clarify:

```
        // singleProcessOOMKill, if set, will prevent the `memory.oom.group` flag from being set for container
	// cgroups in cgroups v2. This causes processes in the container to be OOM killed individually instead of as
	// a group.
```

**Comment by kannon92** [comment]
/hold cancel
LGTM from my end.

**Comment by tzneal** [comment]
pull-kubernetes-node-kubelet-serial-containerd failure looks to be the standard flakes.

**Comment by haircommander** [comment]
new tests passed, others seem like flakes

/lgtm

**Comment by kannon92** [comment]
/assign @mrunalp @dchen1107 
for sig-node approver

Still need someone from api-review to look over the kubelet config.

**Comment by mkarrmann** [comment]
@tzneal As someone whose made 0 contributions to K8s, I just wanted to chime in with my opinion here (won't be offended if it's dismissed out-of-hand :)).

I think that this configuration should be at the container-level, not at the Kubelet. Whether or not a single process or the entire container should be killed is entirely dependent on the specific application, and likely doesn't have much of an impact outside of the specific container in question. So, I think people writing the config for a container are going to be in a far better position to determine what the behavior for their container will be than admins setting the kubelet config.

Of course, nothing here precludes adding the container-level configuration in addition to this configuration (or using `nodeSelector` to accomplish something similar), but I'm worried that having both types of configuration would be unecessarily confusing, and little advantage over just offering container-level config. So, I'd prefer to not see this config option added in the first place.

**Comment by liggitt** [comment]
If this has agreement on direction from kubelet component owners (since there's no linked design, it's not clear to me that it does), get the kubelet packages reviewed / approved, then relabel for api-review for the mechanical flag / config plumbing.

/remove-label api-review

**Comment by iholder101** [comment]
> > this feature is not recommended yet so maybe we can pass on a doc update.
> 
> It's not that the feature isn't recommended yet, I don't think it will ever be recommended. With cgroupsv2, `memory.oom.group` is set by default which causes OOM kills to work like the documentation has stated (i.e. [kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource](https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/) where the docs talk about containers being killed). It didn't work that way previously due needing a user space handler to implement what `memory.oom.group` does automatically.
> 
> This change is just to add support for the hopefully small number of users that want the arguably broken behavior back where individual processes in a container can be OOM killed, but the remaining processes keep running.

Is there a specific use-case that's needed to be supported? If so, what is it?

I kind of miss the point of supporting a broken cgroup v1 behavior for backward-compatibility's sake in a point of time where cgroup v1 is deprecated.

Another thing to mention is that the killed process won't get restarted automatically. Something would need to either restart the process inside the container or restart the container itself when needed.

**Comment by mkarrmann** [comment]
@iholder101 

> Is there a specific use-case that's needed to be supported? If so, what is it?

I think there's lots of situations where not automatically killing the entire container is desirable behavior. The most common example I've seen others point to is postgres, which is designed to let the main postmaster process do clean up in case that a child process crashes. A hard SIGKILL to everything is undesirable from a stability perspective.

A pattern I've used is having a main process on a container which handles light housekeeping logic (e.g. reading from a queue, heartbeating) and child process which does the heavy lifting of handling a request. I've even been forced to use this pattern in Python applications due to GIL-locking issues. In this case, it's generally going to the child process which is the proximate cause for the OOM, and most likely what the kernel is going to target. In this pattern, the main process can detect that the request itself is "bad" because it resulted in a killed process, and can handle the appropriately (e.g. return an error, discard the message, etc.). If the entire container were OOMKilled, then there's no opportunity to do any housekeeping and the rest of the system has no insight into what went wrong (maybe pod disconnected due to a network issue?), meaning I need to retry the job.

As you note, the child process will need to be restarted, but most Process Pool libaries handle this out of the box, so isn't much of an issue.

With all that being said, I stand by my above comment that I don't believe this configuration belongs in the kubelet, since the situations where this is desirable behavior is very application/container-specific. I'd rather have the configuration live at that level, and I'm worried having the configuration at the kubelet level as would muddy the water.

**Comment by gdhagger** [comment]
We have a bunch of processes that currently have a python wrapper program that's managing a task through a bunch of phases that call out to other binaries.  If any of those binaries fails due to a segfault - or more frequently, and OOM event - the python code is responsible for seeing that it OOMed and updating our records of how that job failed.

There are definitely cons to this setup, and more K8s native ways to reimplement it, but it's a big lift to refactor our processes to handle this in the near-medium term.  I'm also certain that we're not the only folks relying on this mode of operation.

**Comment by iholder101** [comment]
Thank you @mkarrmann and @gdhagger for the detailed explanation!

> This change is just to add support for the hopefully small number of users that want the arguably broken behavior

If so, I'd say that it's not a broken behavior, but rather a feature that's needed for exceptional use-cases and that needs to be used with care.

> I think that this configuration should be at the container-level, not at the Kubelet.

Since this behavior is expected to be valuable only for exceptional use-cases, I tend to agree that we shouldn't define it on the kubelet (=node) level, unless there's a use-case for turning it on for the whole node that I can't think of.

**Comment by superbrothers** [comment]
I agree that the container level is best, but if that is the case, I would suggest that https://github.com/kubernetes/kubernetes/pull/117793 be reverted. This change would be fatal to our workloads. https://github.com/kubernetes/kubernetes/pull/117793#issuecomment-1837400726

**Comment by mkarrmann** [comment]
@superbrothers I think the community consensus is that cgroup-aware OOMs are better default behavior. The K8s docs have always been written in a way that implies that this is the case, so arguably it was a bug that this wasn't the default behavior. It also takes forethought to take advantage of non-cgroup-aware OOMKiller, which most devs won't do.

The fact that #117793 broke workloads is unfortuanate, but I also think the damage is done. Now that that's been released, there's likely workloads that now implicitely depend on that behavior, so reverting that behavior would break those.

I'm glad to see we agree that container-level config would be best :) I'm brand new to K8s development, but I have a goal of shipping container-level configuration for this flag. Hopefully that can be completed and out the door relatively soon, so that all use cases can be satisfied.

**Comment by utam0k** [comment]
@mkarrmann Hi, Matt. 
> I think the community consensus is that cgroup-aware OOMs are better default behavior.

I think this is the community consensus.
https://github.com/kubernetes/kubernetes/pull/117793#issuecomment-1845678609

It would take a lot of time and effort to get this configuration into the container level. I fear that many users will not be able to migrate cgroup v1/v2 https://github.com/kubernetes/enhancements/pull/4572 in the meantime.
You may be right about the final landing point, but I just think a practical solution would be to support it once with kubelet, write documentation for migration, and deprecate the kubelet configuration once it is possible to configure it at the container level. What do you think?

**Comment by mkarrmann** [comment]
@utam0k Hello!

I personally am unsure if introducing and documenting a feature which is intended to be deprecated soon is a great idea (I don't know whether or not there's precedent for that in K8s). I also am quite sure that the kubelet solution will be overly burdensome for the majority developers who wish to configure this (I bet in most orgs, the vast majority of devs don't even know who to reach out to to discuss making a kubelet change in their nodes). Upon reflection, I think I actually would prefer temporarily reverting https://github.com/kubernetes/kubernetes/pull/117793, as opposed to introducing this config to compensate.

But that's all just my personal opinion and I don't have much conviction on this point—I'm definitely not going to object to anything trying to unblock users. I also grant your point that shipping the container-level config simply may take too long.

**Comment by marquiz** [comment]
> I think that this configuration should be at the container-level, not at the Kubelet. Whether or not a single process or the entire container should be killed is entirely dependent on the specific application

I agree that this should not be a kubelet/node level setting. I also feel that one problem with the kubelet flags is that the users don't have good visibility and in worst case nodes might be configured differently.

A possible solution to provide per-container control would be [QoS-resources (KEP)](https://github.com/kubernetes/enhancements/pull/3004). The OOM kill behavior would be a Kubernetes-managed QoS-resources with two possible settings (classes): group kill or process kill. It would also have the advantage to set per-cluster/per-node/per-namespace defaults. Plus the user could always see what berhavior their applications get.

**Comment by yujuhong** [comment]
I think we need to decide whether this is a behavior/setting we want to support long-term, or simply a temporary crutch to help with the transition from cgroup v1 to v2. If it's the latter, we should not be adding anything to the k8s core API.

**Comment by mkarrmann** [comment]
I've opened an issue describing my proposal: https://github.com/kubernetes/kubernetes/issues/124253

I tried to thoroughly articulate why it's important that first-class support for this behavior is provided.

**Comment by utam0k** [comment]
@tzneal  @haircommander @mrunalp @dchen1107
Sorry for pinging you, but I'd be happy to know the status of this PR in the sig-node team. I think it is a bit risky to enter this transitional period between cgroup v1 and v2 with this PR in limbo.
Two users[^1][^2] have already struggled with this behavior. I guess more users who will probably migrate to cgroup v2 soon. For example, it is possible that Ubuntu 24.04 was released and they will try it together to migrate v1 to v2.

[^1]: https://github.com/kubernetes/kubernetes/pull/117793#issuecomment-2059012668
[^2]: https://github.com/kubernetes/kubernetes/pull/117793#issuecomment-1837400726

**Comment by mkarrmann** [comment]
For the record, I largely retract my original comment about not wanting this merged. It will take a while for a proper solution to be merged, and I agree this is important that it deserves a sooner fix (even if this solution will be inconvenient for many).

One suggestion @tzneal :

Make the config of type `*bool` with default `nil`, where the semantics of `nil` is that no guarantees will be made at the kubelet level which option will be used. In practice, this would mean respect any Container/Pod-level configuration that is added later, but the kubelet can still default to using group OOM kills.

I still think it's important to eventually have an option for configuring this per Pod in the K8s API, in which case it'd be best to avoid confusion or ambiguity regarding how that will interact with this feature.

**Comment by utam0k** [comment]
@mkarrmann Thanks for your understanding and effort 🙏 I totally agree with you.
> For the record, I largely retract my original comment about not wanting this merged. It will take a while for a proper solution to be merged, and I agree this is important that it deserves a sooner fix (even if this solution will be inconvenient for many).

**Comment by tzneal** [comment]
Closing this for now as I don't have time at the moment.

**Comment by utam0k** [comment]
@tzneal Thanks for your effort 🙏 I understand your situation. Can I take over this issue?

**Comment by tzneal** [comment]
> @tzneal Thanks for your effort 🙏 I understand your situation. Can I take over this PR?

Sure thing, please do!

**Pull request closed without merging**